spec/lib/food_robot_spec.rb

```diff
@@ -0,0 +1,294 @@
+      end
+
+      context 'with a non-functional schedule chain' do
+        let!(:chain) { create(:schedule_chain, day_of_week: monday.wday) }
+        let!(:recipient) { create(:recipient_schedule, schedule_chain: chain, position: 1) }
+        let!(:donor) { create(:donation_schedule, schedule_chain: chain, position: 2) }
+
+        subject { described_class.generate_log_entries(monday) }
+
+        it 'generates no logs' do
+          expect { subject }.not_to change { Log.count }
+        end
+      end
+
+      context 'with a weekly schedule chain' do
+        let!(:chain) { create(:schedule_chain, day_of_week: monday.wday, num_volunteers: 3) }
+        let!(:donor1) { create(:donation_schedule, schedule_chain: chain, position: 1) }
+        let!(:donor2) { create(:donation_schedule, schedule_chain: chain, position: 2) }
+        let!(:recipient) { create(:recipient_schedule, schedule_chain: chain, position: 3) }
+
+        context 'for a day other than the day of the schedule chain' do
+          subject { described_class.generate_log_entries(tuesday) }
+
+          it 'generates no logs' do
+            expect { subject }.not_to change { Log.count }
+          end
+        end
+
+        context 'for the day of the schedule chain' do
+          subject { described_class.generate_log_entries(monday) }
+
+          it 'generates a log for each donor' do
+            expect { subject }.to change { Log.count }.by(2)
+          end
+
+          it 'returns the number of created logs' do
+            n, _ = subject
+            expect(n).to eq(2)
+          end
+
+          it 'returns the number of skipped logs' do
+            _, n_skipped = subject
+            expect(n_skipped).to eq(0)
+          end
+
+          context 'with a log already generated' do
+            before do
+              create(
+                :log,
+                when: monday,
+                schedule_chain: chain,
+                donor: donor1.location,
+                recipients: [recipient.location]
+              )
+            end
+
+            it 'does not generate duplicate logs' do
+              expect { subject }.to change { Log.count }.by(1)
+            end
+
+            it 'returns the number of created logs' do
+              n, _ = subject
+              expect(n).to eq(1)
+            end
+
+            it 'returns the number of skipped logs' do
+              _, n_skipped = subject
+              expect(n_skipped).to eq(1)
+            end
+          end
+        end
+      end
+
+      context 'with a schedule chain ending at a hub' do
+        let!(:chain) { create(:schedule_chain, day_of_week: monday.wday, num_volunteers: 3) }
+        let!(:donor) { create(:donation_schedule, schedule_chain: chain, position: 1) }
+        let!(:hub) { create(:schedule, schedule_chain: chain, location: create(:hub), position: 2) }
+
+        subject { described_class.generate_log_entries(monday) }
+
+        it 'does not generate a log for the hub' do
```

Comment:
I think it's ok not to in this case.

A "hub" can be either a donor or a recipient based on where it is in the `ScheduleChain`. If it's the last stop in the chain, it's considered a recipient. Otherwise, it's considered a donor. This test is checking to make sure that our `generate_log_entries` method does not consider the hub at the end of the chain to be a donor, and that it does not create a separate log for it. Without the last condition on [this line](https://github.com/boulder-food-rescue/food-rescue-robot/blob/master/lib/food_robot.rb#L61), `Log.count` will change by 2 instead of 1.